### PR TITLE
Fork instructions in README

### LINKED ISSUE #18: Instructions for setting up a fork
Hi @jessieay !

Just some notes on setting up a fork and a request that we cover it in the README. (There's a lot of coverage in the README now of how it works, and that's great, but nothing AFAIK about setting up another sudo-sandwich instance.)

Things I ran into (easily fixed):
- Set SLUG and PASSWORD env vars to what is in the manifest
- Set the secret for attr_encrypted in `ENV['ENCRYPTION_KEY']` -- need guidance as to size / how to generate it "correctly", though -- I used `SecureRandom.hex(32)`.
- SSO didn't work -- I had to set `ENV['SSO_SALT']` to what my generated manifest had for `"sso_salt"`.

Other notes (probably should have opened another Issue, sorry):
- We're logging all parameters and we probably want to filter them. Although, for dev usage, this is helpful to know what was sent.
- Catch some errors on provisioning failure and return a helpful JSON error rather than nothing. the Failed Attempts table didn't help me with any of the above, and I had to go into the Heroku app logs.

### PR DESCRIPTION
**Pull request by mathias**
Fixes #18.

Documents how to set up a new fork of Sudo Sandwich on Heroku and moves Set up instructions to the README.

[Rendered](https://github.com/heroku/sudo-sandwich/blob/mathias-fork-instructions-in-readme/README.md#creating-a-fork-on-heroku)

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -312,3 +312,62 @@ reported properly.
 
 *NOTE*: the current implementation does not handle all edge cases at this time.
 If you send invalid data, it is possible for it to silently fail.
+
+## Creating a fork on Heroku
```
Comment (jessieay) [inline README.md]: Since forking has many meanings, maybe a more specific title would be "How to fork this add-on in order to create your own Heroku add-on"

File: README.md [hunk 2]
```diff
@@ -312,3 +312,62 @@ reported properly.
...
+
+If you want to run your own add-on based on this codebase, you should follow these instructions. These instructions are for deploying your add-on to Heroku, which is not required. You may have to modify the deploy instructions for other environments.
+
+To start, you may wish to fork the codebase to your own private GitHub repo, so that you can make changes for your specific add-on's use case.
```
Comment (jessieay) [inline README.md]: I think we can start with

> To start, fork the codebase to your own private GitHub repo,

and perhaps link to a GH help article on forking?

File: README.md [hunk 3]
```diff
@@ -312,3 +312,62 @@ reported properly.
...
+
+Create a Heroku app. Typically, we name the Heroku app after the slug (or commandline identifier) for the add-on. If I was creating an add-on with the slug of `sudo-sandwich`, I'd probably name the Heroku app `sudo-sandwich` as well. This is not required; and you can name it whatever you'd like. Please note that `sudo-sandwich` is already taken, and slugs are both immutable and must be globally unique!
```
Comment (jessieay) [inline README.md]: I think commandline is usually 2 words or separate with a hyphen

File: README.md [hunk 4]
```diff
@@ -312,3 +312,62 @@ reported properly.
...
+
+```
+heroku apps:create sudo-sandwich
```
Comment (jessieay) [inline README.md]: someone will need the heroku CLI for this, perhaps mention in previous paragraph?

File: README.md [hunk 5]
```diff
@@ -312,3 +312,62 @@ reported properly.
...
+```
+heroku addons:create heroku-postgresql --app sudo-sandwich
+```
+
+**TODO** Instructions for deploy the code to Heroku go here.
+
+Generate a manifest with the [addons-admin](https://github.com/heroku/heroku-cli-addons-admin) CLI plugin. See the linked repo for install instructions. After install, run `heroku addons:admin:manifest:generate` and follow the prompts.
+
+
+Allow the `addons-admin` CLI plugin to generate a secret and SSO salt for you. It will save a new file called `addon-manifest.json`. You may need to edit this file to suit your local dev environment's port.
```
Comment (jessieay) [inline README.md]: > You may need to edit this file to suit your local dev environment's port.

can you be more specific about which part of the file I'd need to edit?

File: README.md [hunk 6]
```diff
@@ -312,3 +312,62 @@ reported properly.
...
+Set secrets from the `addon-manifest.json` on your newly-created app under the following config vars:
+
+```
+heroku config:set SLUG=<slug-from-manifest> PASSWORD=<password-from-manifest> SSO_SALT=<salt-from-manifest> --app sudo-sandwich
+```
+
+You'll also have to generate an encryption key for the database to store secrets with. In a Ruby terminal (pry or irb), run `require 'securerandom'; SecureRandom.hex(32)` to generate an encryption key of the appropriate length. Then set that key on the Heroku app:
+
+```
+heroku config:set ENCRYPTION_KEY=<value-from-securerandom> --app sudo-sandwich
```
Comment (jessieay) [inline README.md]: does this need to bet set for the dev or test env at all?

File: README.md [hunk 7]
```diff
@@ -312,3 +312,62 @@ reported properly.
...
+```
+
+Once you've completed the above, you should be ready to push the manifest to the Heroku API:
+
+```
+heroku addons:admin:manifest:push
+```
+
+If that doesn't work, you'll need to sign up at the [Partner Portal](https://addons-next.heroku.com) as an Add-on Partner.
```
Comment (jessieay) [inline README.md]: do I retry after signing up?

File: README.md [hunk 8]
```diff
@@ -312,3 +312,62 @@ reported properly.
...
+Create a Heroku app. Typically, we name the Heroku app after the slug (or commandline identifier) for the add-on. If I was creating an add-on with the slug of `sudo-sandwich`, I'd probably name the Heroku app `sudo-sandwich` as well. This is not required; and you can name it whatever you'd like. Please note that `sudo-sandwich` is already taken, and slugs are both immutable and must be globally unique!
+
+```
+heroku apps:create sudo-sandwich
+```
+
+You'll need a Heroku Postgresql database to store data in:
+
+```
+heroku addons:create heroku-postgresql --app sudo-sandwich
```
Comment (jessieay) [inline README.md]: maybe replace `sudo-sandwich` with `your-app-slug` ?

File: README.md [hunk 9]
```diff
@@ -312,3 +312,62 @@ reported properly.
...
+```
+heroku addons:admin:manifest:push
+```
+
+If that doesn't work, you'll need to sign up at the [Partner Portal](https://addons-next.heroku.com) as an Add-on Partner.
+
+You will also want to scale the `worker` dyno up so that async provisioning can be handled in the background:
+
+```
+heroku ps:scale worker=1 --app sudo-sandwich
```
Comment (jessieay) [inline README.md]: maybe replace `sudo-sandwich` with `your-app-slug` ?

File: README.md [hunk 10]
```diff
@@ -312,3 +312,62 @@ reported properly.
...
+```
+
+You will need to go to the [Partner Portal](https://addons-next.heroku.com) and add [plans](https://github.com/heroku/sudo-sandwich/blob/master/app/models/sandwich.rb#L2-L8) to your add-on based on the hardcoded values in the `Sandwich` class. You may wish to modify these, for example, to create a plan called `async` for testing async provisioning.
```
Comment (jessieay) [inline README.md]: Maybe add a caveat here: if someone wants to use the same plans, the will need to do this. They just need to make sure that the pans in the partner portal match the `Sandwich` plans

File: README.md (hunk 6 above)
Comment (mathias) [reply README.md]: I guess we're missing the setup for dev here. I'll point them at .env.sample and `bin/setup`.

Comment (mathias) [reply README.md]: Fixed in 10f13cafe6203c90c357c8ff9f0abffe91d6d207.

**Comment by mathias** [comment]
This should be ready for review again @jessieay .

File: README.md [hunk 11]
```diff
@@ -312,3 +360,75 @@ reported properly.
 
 *NOTE*: the current implementation does not handle all edge cases at this time.
 If you send invalid data, it is possible for it to silently fail.
+
+## How to fork this add-on to create your own SudoSandwich instance on Heroku
```
Comment (jessieay) [inline README.md]: `SudoSandwich` -> `Sudo Sandwich`

File: README.md [hunk 12]
```diff
@@ -312,3 +360,75 @@ reported properly.
...
+
+If you want to run your own add-on based on this codebase, you should follow these instructions. These instructions are for deploying your add-on to Heroku, which is not required. You may have to modify the deploy instructions for other environments.
```
Comment (jessieay) [inline README.md]: can do we do line breaks in these paragraphs for easier editing / inline commenting? my editor has an easy way to do this. if yours does not we can always do it in a separate PR.

File: README.md [hunk 13]
```diff
@@ -312,3 +360,75 @@ reported properly.
...
+
+To start, you may wish to fork the codebase to your own private GitHub repo, so that you can make changes for your specific add-on's use case. If you intend to do local development, see [Set up](#set-up).
+
+Create a Heroku app. Typically, we name the Heroku app after the slug (or command line identifier) for the add-on. If I was creating an add-on with the slug of `sudo-sandwich`, I'd probably name the Heroku app `sudo-sandwich` as well. This is not required; and you can name it whatever you'd like. Please note that `sudo-sandwich` is already taken, and slugs are both immutable and must be globally unique!
```
Comment (jessieay) [inline README.md]: 🥇

File: README.md [hunk 14]
```diff
@@ -312,3 +360,75 @@ reported properly.
...
+heroku addons:create heroku-postgresql --app your-app-slug
+```
+
+You will also want to scale the `worker` dyno up so that async provisioning can be handled in the background:
+
+```
+heroku ps:scale worker=1 --app your-app-slug
+```
+
+Deploying the code to Heroku either [using git](https://devcenter.heroku.com/articles/git) or [the Heroku Dashboard with the GitHub integration](https://devcenter.heroku.com/articles/github-integration).
```
Comment (jessieay) [inline README.md]: `Deploying` -> `Deploy`

File: README.md [hunk 15]
```diff
@@ -312,3 +360,75 @@ reported properly.
...
+
+Generate a manifest with the [addons-admin](https://github.com/heroku/heroku-cli-addons-admin) CLI plugin. (See the linked repo for plugin install instructions.) After install, run `heroku addons:admin:manifest:generate` and follow the prompts.
```
Comment (jessieay) [inline README.md]: seems like in this section you are doing inline markdown links but above you were listing links below. Not a big deal but it might be nice to be internally consistent on which style. I don't have a strong preference either way

File: README.md [hunk 16]
```diff
@@ -312,3 +360,75 @@ reported properly.
...
+```json
+    "production": {
+      "sso_url": "https://your-app-slug.herokuapp.com/sso/login",
+      "base_url": "https://your-app-slug.herokuapp.com/heroku/resources"
+    },
+```
+
+You may also need to edit this file to suit your local dev environment's port, inside the `"api.test.sso_url"` and `"api.test.base_url"` keys. For more details on editing this file, see the [Manifest](https://devcenter.heroku.com/articles/add-on-manifest) docs.
+
+Set secrets from the `addon-manifest.json` on your newly-created app under the following config vars:
```
Comment (jessieay) [inline README.md]: this is a very helpful step!

File: README.md (hunk 12 above)
Comment (jessieay) [reply README.md]: I usually just let me editor to the newlines for me so I usually do the first style but I can see why you might prefer the latter. Fine with either!

File: README.md (hunk 15 above)
Comment (jessieay) [reply README.md]: oh was the setup guide existing? This diff makes it look like you added, for example, `[Ruby on Rails]: http://rubyonrails.org`

Comment (mathias) [reply README.md]: Ah right, that content came out of `CONTRIBUTING.md` which I have not changed/removed yet.

I didn't get an answer for a `CONTRIBUTING.md` we use for all @heroku open source repos, so we'll have to put something basic in there for now.

Sorry, I'd totally forgotten about that! Changes were made too long ago.

File: README.md (hunk 16 above)
Comment (mathias) [reply README.md]: 👍

File: README.md (hunk 15 above)
Comment (jessieay) [reply README.md]: no biggie! happy to keep as-is for now if you just want to :shipit: 

this doc is great

Comment (mathias) [reply README.md]: Cool. I'd like to keep improving it, but it will take more iterations :)
